keps/1714-fair-sharing/README.md

```diff
@@ -0,0 +1,366 @@
+important) to ensure equal extra space (irregardless of their given quota) for both teams. 
+
+
+#### Story 2
+I have two departments in my company D1 and D2 (modeled as Cohorts, under one
+top-level Cohort). Each of the departments has 3 teams: A, B, C. Each of the teams
+submits their workloads to the appropriate CQ: 1A, 1B, 1C, 2A, 2B, 2C. Each of the
+teams/CQs is given some guaranteed resources within the big ML cluster. Each team
+is also allowed to borrow as many resources from other teams as they want, as long
+as they are not needed by the owner. However, due to internal accounting, company
+internal politics, etc, free resources should first be distributed within the
+department. For example:
+
+* if CQ 1A has free resources, and both 1B and 1C are in need of extra resources 
+to run their workloads, then 2A, 2B and 2C should not be given a share of 1A,
+that would block any workload from 1B or 1C from running. 
+
+* If CQ 1A has free resources and 1B and 1C do not need any of those to start
+a workload (there are no pending workloads or they are too big anyway) then 2A, 2B,
+2C should be given an equal share of the resources that correspond to their needs.
+
+#### Story 3
+I have 3 cohorts inside my organization (top cohort): CS, C1 and C2. CS contains
+a single CQ that is given 300 GPU nominal quota. C1 contains 3 CQ: 1A, 1B and 1C,
+among which 1C is considered to be 3 times more important than others. C2 contains
+only one CQ: 2A.  Each of the queues 1A, 1B, 1C and 2C has a long backlog of jobs
+to execute. I want to distribute the resources from CS in the following way:
+
+* Both C1 and C2 get 150 GPUs each, because they are considered equally important.
+* Inside C1, out of these 150 GPUs, 1A gets 30, 1B gets 30 and 1C gets 90 GPUs. 
+1A and 1B are equally important but 1C is 3x more important than 1A and 1B.
+* Inside C2 there is just a single CQ, so it gets all 150 GPUS.
+
+### Risks and Mitigations
+
+* Fair sharing may increase the number of preemptions in the system vs current
+state where the first workload to acquire unused resources keeps them until it
+finishes. Mitigations include:
+    * Introduce minimum execution time before workloads can be preempted for fair sharing.
+    * Introduce delayed fair share enforcement - new workloads have to wait a bit before preempting others to get their share.
+
+* Fair sharing may decrease utilization of the unused resources while attempting to
+distribute them fairly, vs provide the tighties bin-packing. To avoid these scenarios,
+users should prefer to run massive workloads under nominal quotas.
+
+
+## Design Details
+
+The enforcement of fair-share revolves around the idea of a share value per CQ or
+Cohort and mechanisms to keep the share values among siblings in the hierarchy to
+be as equal as possible. This is achieved via two mechanisms:
+
+* During admission, prioritize admitting workload from the ClusterQueues and cohorts that have the lowest share value.
```

Comment:
One general question is: since DRF will try to distribute the workloads to different resource pool(CQ) as a result of favoring the smaller share value, but preemption will try to evict the borrowers, isn't this conflict, which means we'll evict more than we're today.

Reply:
There is no conflict. See the rules S2-a and S2-b in `Choosing workloads from ClusterQueues for preemption`.

It's easy to see that S2-a increases the value of the smallest CQ but it remains being the one with smallest share.

S2-b has a typo, but see my comment below.

That said, fair sharing will indeed increase the number of preemptions overall, but the algorithm should reach a steady state. In other words, there wouldn't be flapping.